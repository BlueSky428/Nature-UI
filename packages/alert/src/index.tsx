import * as React from 'react';
import clx from 'clsx';
import { FiInfo, FiAlertCircle, FiAlertTriangle, FiCheckCircle } from 'react-icons/fi';

const SUBTLE_TEXT = 'text-gray-800';

export const ALERT_STATUSES = {
  error: {
    bg: 'bg-red-200',
    iconColor: 'text-red-600 mr-3',
    icon: FiAlertCircle,
    variant: {
      solid: 'bg-red-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  info: {
    bg: 'bg-blue-200',
    iconColor: 'text-blue-600 mr-3',
    icon: FiInfo,
    variant: {
      solid: 'bg-blue-600',
      subtle: SUBTLE_TEXT,
    },
  },
  success: {
    bg: 'bg-green-200',
    iconColor: 'text-green-600 mr-3',
    icon: FiCheckCircle,
    variant: {
      solid: 'bg-green-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  warning: {
    bg: 'bg-orange-200',
    iconColor: 'text-orange-600 mr-3',
    icon: FiAlertTriangle,
    variant: {
      solid: 'bg-orange-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
};

interface AlertOptions {
  /**
   * The status of the alert
   */
  status?: keyof typeof ALERT_STATUSES;
  /**
   * The variant of the alert style to use
   */
  variant?: 'solid' | 'subtle';

  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;

  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ElementType;
}

export type AlertProps = AlertOptions;

const BASE_STYLE = 'px-4 py-3 flex items-center';

const Alert: React.ForwardRefExoticComponent<AlertProps> = React.forwardRef(
  (props: AlertProps, ref: React.Ref<any>) => {
    const {
      className = '',
      status = 'success',
      children,
      variant = 'subtle',
      component: Component = 'div',
    } = props;

    // const Component = 'div';

    const VARIANT: string = ALERT_STATUSES[status].variant[variant];

    const componentClass = clx(BASE_STYLE, {
      [className]: className,
      [ALERT_STATUSES[status].bg]: status,
      [VARIANT]: variant,
    });

    const Icon = ALERT_STATUSES[status];
    const iconClasses = clx({
      [VARIANT]: variant,
      'mr-3': variant,
      [Icon.iconColor]: variant !== 'solid',
    });

    const IconComponent = Icon.icon;

    return (
      <Component className={componentClass} ref={ref}>
        {status && <IconComponent className={iconClasses} size={20} />}
        <span>{children ? children : 'This is an alert ðŸ™‚'}</span>
      </Component>
    );
  }
);

Alert.displayName = 'Alert';

export default Alert;
