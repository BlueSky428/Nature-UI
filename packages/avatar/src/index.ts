export * from './avatar';
export * from './avatarGroup';
