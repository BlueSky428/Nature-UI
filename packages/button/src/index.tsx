import * as React from 'react';
import clx from 'clsx';

interface ButtonProps {
  /**
   * The label of the button
   */
  'aria-label'?: string;
  /**
   * The text color of the button. Use a color key passed in theme.colors.
   */
  text?: string;
  /**
   * The background color of the button. Use a color key passed in theme.colors.
   */
  color?: string;
  /**
   * The variant of the button style to use.
   */
  variant?: 'outline' | 'ghost' | 'link' | 'solid' | 'none';
  /**
   * If true, the button will be disabled.
   */
  isDisabled?: boolean;
  /**
   * If true, the button will show a spinner.
   */
  isLoading?: boolean;
  /**
   * The label to show in the button when isLoading is true. If no text is passed, it only shows the spinner
   */
  loadingText?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'none';
  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;
  /**
   * The lightness or how dark you want the background color to be
   */
  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   * defaults to "button"
   */
  component?: React.ElementType;
  id?: string;
  /**
   *  when the href prop is set it changes the tag name to a
   */
  href?: string;
  onClick?: () => any;
}

const darken = (value: string): string => {
  const splitStr = value.split('-');
  const dept = Number(splitStr[1]);
  const color = splitStr[0];

  if (dept >= 800) {
    return `${color}-${dept - 200}`;
  }

  return `${color}-${dept + 200}`;
};

const lighten = (value: string): string => {
  const color = value.split('-')[0];

  return `${color}-${100}`;
};

const Button = React.forwardRef((props: ButtonProps, ref: React.Ref<any>) => {
  const {
    component: Component = 'button',
    variant = 'solid',
    color = 'teal-500',
    text = 'white',
    size = 'md',
    children,
    ...rest
  } = props;
  const DEFAULT_CLASS = 'block focus:shadow-outline focus:outline-none rounded font-semibold flex';
  const STYLES = {
    solid: `bg-${color} text-${text} hover:bg-${darken(color)}`,
    outline: `bg-transparent hover:bg-${lighten(
      text
    )} text-${text} border border-${text} focus:border-transparent`,
    ghost: `hover:bg-${lighten(text)} text-${text}`,
    link: `hover:underline text-${text}`,
    md: 'px-4 py-2',
  };

  let BTNClass: string;

  if (variant === 'none') {
    BTNClass = '';
  } else {
    BTNClass = clx(DEFAULT_CLASS, {
      [STYLES[size]]: size && variant !== 'link',
      [STYLES[variant]]: variant,
    });
  }

  return (
    <Component className={BTNClass} ref={ref} {...rest}>
      {children ? children : 'Button'}
    </Component>
  );
});

export default Button;
