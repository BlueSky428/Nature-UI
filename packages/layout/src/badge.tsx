import { clsx, forwardRef, nature, PropsOf } from '@nature-ui/system';
import { __DEV__ } from '@nature-ui/utils';

export type BadgeProps = PropsOf<typeof nature.span>;

export type BadgeType = BadgeProps & {
  color?: string;
  variant?: 'solid' | 'subtle' | 'outline';
  css?: any;
};

export const Badge = forwardRef<BadgeType, 'span'>((props, ref) => {
  const { children, className, css, ...rest } = props;

  const DEFAULTS =
    'rounded-sm px-1 py-0.5 text-xs uppercase font-semibold inline-block align-middle';

  const _classNames = clsx(DEFAULTS, className);

  const values = {
    ref,
    className: _classNames,
    ...rest,
  };

  return (
    <nature.span css={{ ...css, lineHeight: 'normal' }} {...values}>
      {children}
    </nature.span>
  );
});

if (__DEV__) {
  Badge.displayName = 'Badge';
}
