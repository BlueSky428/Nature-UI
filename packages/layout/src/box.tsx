import * as React from 'react';
import { nature, PropsOf, forwardRef } from '@nature-ui/system';

const BoxLayout = nature('div');

export type BoxProps = PropsOf<typeof BoxLayout>;

// type Omitted = "size" | "boxSize" | "width" | "height" | "w" | "h"

export type SquareProps = BoxProps & {
  /**
   * The size (width and height) of the square
   * It uses the range value in your tailwind.config.js file
   * i.e lg = w-2/3
   */
  size?: number | 'xs' | 'sm' | 'md' | 'lg' | 'xl';
  /**
   * If `true`, the content will be centered in the square
   */
  centerContent?: boolean;
  /**
   * If variant is set to circle then the @borderRadius will
   * take the maximum borderRadius in your tailwind.config.js file.
   */
  variant?: 'square' | 'circle';
};

const sizes = {
  xs: 'w-1/6',
  sm: 'w-1/3',
  md: 'w-2/4',
  lg: 'w-2/3',
  xl: 'w-11/12',
};

export const Box = forwardRef<SquareProps>((props, ref) => {
  const { children, ...rest } = props;

  const values = {
    ref,
  };

  return (
    <BoxLayout {...rest} {...values}>
      {children}
    </BoxLayout>
  );
});
