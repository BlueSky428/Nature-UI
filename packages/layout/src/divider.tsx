import * as React from 'react';
import { forwardRef, nature, PropsOf } from '@nature-ui/system';
import clsx from 'clsx';

const DividerElem = nature('hr');

export type DividerProp = Omit<PropsOf<typeof DividerElem>, 'children'> & {
  /**
   * The thickness of the divider
   */
  size?: number;
};

export const Divider = forwardRef<DividerProp>((props, ref) => {
  const { className = '', size = 0, color = 'gray-200', ...rest } = props;

  const _className = clsx({
    [`border border-${size}`]: size > 0,
    [`border-${color}`]: color,
    [className]: className,
  });

  return <DividerElem className={_className} {...rest} ref={ref} />;
});
