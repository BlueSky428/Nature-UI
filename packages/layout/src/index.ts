export * from './box';
export * from './stack';
export * from './divider';
export * from './badge';
