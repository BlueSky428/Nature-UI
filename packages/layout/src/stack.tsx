import * as React from 'react';
import { forwardRef, nature, PropsOf, clsx } from '@nature-ui/system';
import { getValidChildren, __DEV__ } from '@nature-ui/utils';
import { css } from 'emotion';

import { Box } from './box';

const StackElem = nature('div');

export type StackProps = PropsOf<typeof StackElem> & {
  /**
   * The space between each stack item
   */
  spacing?: string | number;
  /**
   * The direction to stack the items.
   */
  direction?: 'row' | 'col';
};

export const Stack = forwardRef<StackProps>((props, ref) => {
  const {
    children,
    spacing = 4,
    direction = 'col',
    className = '',
    ...rest
  } = props;

  const validChildren = getValidChildren(children);

  const clones = validChildren.map((child, index) => {
    const isLast = index + 1 === validChildren.length;

    const SPACING = typeof spacing === 'number' ? `${spacing}px` : spacing;

    const ROW = css`
      margin-right: ${SPACING};
    `;

    const COL = css`
      margin-bottom: ${SPACING};
    `;

    if (!isLast) {
      const _className = clsx({
        [COL]: spacing && direction === 'col',
        [ROW]: spacing && direction === 'row',
      });

      return (
        <Box key={Number(index)} className={_className}>
          {child}
        </Box>
      );
    }

    return <Box key={Number(index)}>{child}</Box>;
  });

  const DEFAULT_CLASS = `flex`;
  const _className = clsx(DEFAULT_CLASS, {
    [`flex-${direction}`]: direction,
    [className]: className,
  });

  return (
    <StackElem className={_className} {...rest} ref={ref}>
      {clones}
    </StackElem>
  );
});

if (__DEV__) {
  Stack.displayName = 'Stack';
}
