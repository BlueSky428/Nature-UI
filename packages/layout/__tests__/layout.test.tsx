import { render } from '@nature-ui/test-utils';
import { Badge, Box } from '../src';

describe('@nature-ui/layout', () => {
  describe('Box', () => {
    test('should render box correctly', () => {
      const tools = render(<Box>This is box</Box>);

      expect(tools.asFragment()).toMatchSnapshot();
    });

    test('as - prop works correctly', () => {
      const tools = render(
        <Box as='a' href='www.google.com'>
          This is box
        </Box>,
      );

      expect(tools.asFragment()).toMatchSnapshot();
    });
  });

  describe('Badge', () => {
    test('should render with default theming', () => {
      const tools = render(<Badge>Badge</Badge>);

      expect(tools.asFragment()).toMatchSnapshot();
    });

    test('override the theming props', () => {
      const tools = render(
        <Badge variant='outline' color='pink-500'>
          Badge
        </Badge>,
      );

      expect(tools.asFragment()).toMatchSnapshot();
    });
  });
});
