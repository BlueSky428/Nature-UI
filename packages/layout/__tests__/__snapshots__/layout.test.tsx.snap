// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nature-ui/layout Badge override the theming props 1`] = `
<DocumentFragment>
  <span
    class="rounded-sm px-1 py-0.5 text-pink-500 uppercase font-semibold inline-block align-middle border-pink-500 css-1kvi9rf-Badge"
  >
    Badge
  </span>
</DocumentFragment>
`;

exports[`@nature-ui/layout Badge should render with default theming 1`] = `
<DocumentFragment>
  <span
    class="rounded-sm px-1 py-0.5 text-gray-700 uppercase font-semibold inline-block align-middle bg-gray-200 css-1kvi9rf-Badge"
  >
    Badge
  </span>
</DocumentFragment>
`;

exports[`@nature-ui/layout Box as - prop works correctly 1`] = `
<DocumentFragment>
  <a
    class=""
    href="www.google.com"
  >
    This is box
  </a>
</DocumentFragment>
`;

exports[`@nature-ui/layout Box should render box correctly 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    This is box
  </div>
</DocumentFragment>
`;
